worker/s3/authentication.go

```diff
@@ -123,33 +125,53 @@ func (b *authenticatedBackend) reloadV4Keys(ctx context.Context) error {
 	return nil
 }
 
-func (b *authenticatedBackend) AuthenticateRequest(w http.ResponseWriter, rq *http.Request, bucket string) bool {
-	perms := noAccessPerms
-	if rq.Header.Get("Authorization") == "" && bucket != "" {
-		// No auth header, we continue without permissions. Request might
-		// still succeed due to bucket policy
-	} else if err := b.reloadV4Keys(rq.Context()); err != nil {
+func (b *authenticatedBackend) AuthenticationMiddleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, rq *http.Request) {
+		// start with no permissions
+		perms := noAccessPerms
+
+		if rq.Header.Get("Authorization") != "" {
+			// auth header found, refresh keys
+			if err := b.reloadV4Keys(rq.Context()); err != nil {
+				writeResponse(w, signature.APIError{
+					Code:           string(gofakes3.ErrInternal),
+					Description:    fmt.Sprintf("failed to reload v4 keys: %v", err),
+					HTTPStatusCode: http.StatusInternalServerError,
+				})
+				return
+			}
+			// verify signature
+			if _, result := signature.V4SignVerify(rq); result != signature.ErrNone {
+				// authentication attempted but failed.
+				writeResponse(w, signature.GetAPIError(result))
+				return
+			}
+			// authenticated request successfully
+			perms = rootPerms
+		}
+
+		// add permissions to context
+		h.ServeHTTP(w, rq.WithContext(context.WithValue(rq.Context(), permissionKey, &perms)))
+	})
+}
+
+func (b *authenticatedBackend) IsAuthenticated(w http.ResponseWriter, rq *http.Request, bucket string) bool {
+	// fetch permissions from context
+	perms := rq.Context().Value(permissionKey).(*permissions)
+
+	// if no authentication has happened so far and no bucket is specified, deny
+	// access
+	if !perms.Authenticated && bucket == "" {
```

Comment:
So as soon as `bucket` is not an empty string `IsAuthenticated` returns true?

Reply:
Yeah. If the bucket is set, there is a chance that the bucket policy still gives you access. But only to specific routes. 
Not sure we can just return `perms.Authenticated` since we need to check its value to decide whether to return an error.

Reply:
I meant the `return true` at the end could/should return `perms.Authenticated` - so after the error check.

Reply:
I can't because then the bucket policies no longer work.